### PR TITLE
Fedora build files

### PR DESCRIPTION
**Pull request by orbisvicis**
Details in notes.md

### CODE REVIEW

File: fedora/copyright [hunk 1]
```diff
@@ -0,0 +1,21 @@
+Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: sonarr
+Source: https://github.com/Sonarr/Sonarr
+
+Files: *
+Copyright: 2010-2014 Sonarr <hello@sonarr.tv>
```
Comment (kayone) [inline fedora/copyright]: prob 2015 now ;) :fireworks:

File: fedora/sonarr.spec [hunk 2]
```diff
@@ -0,0 +1,100 @@
+%define sonarr_url_service  http://services.nzbdrone.com/v1/update/master
```
Comment (kayone) [inline fedora/sonarr.spec]: http://services.sonarr.tv/v1/update/master

File: fedora/sonarr.spec [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+License:        GPLv3+
+URL:            https://sonarr.tv/
+Source0:        %{sonarr_url_package}
+Source1:        copyright
+Source2:        license
+Source3:        sonarr.systemd
+
+Requires:           sqlite >= 3.7
+Requires:           mediainfo >= 0.7.52
+Requires:           mono-complete >= 3.10
```
Comment (kayone) [inline fedora/sonarr.spec]: should prob follow the same dependencies as https://github.com/Sonarr/Sonarr/blob/develop/debian/control

**Comment by kayone** [comment]
Thank you for this! definitely something we want to get included. couple of questions,

- Is there something Fedora specific here? or can we just make this an rpm thing rather than Fedora?

**Comment by kayone** [comment]
We would ideally want to have our own rpm repo just like we do for debian, it would be great if you could give us some help with that.

File: fedora/sonarr.spec (hunk 3 above)
Comment (orbisvicis) [inline fedora/sonarr.spec]: For the most part it does, except:

* libmono-cil-dev has no fedora counterpart. The packages available are @ [Xamarin Mono Repo](http://download.mono-project.com/repo/centos/RPMS/).
* sonarr's wiki recommends mono 3.10, I think this is a good idea, and it's the only mono-3 branch available to fedora.

**Comment by orbisvicis** [comment]
> Is there something Fedora specific here? or can we just make this an rpm thing rather than Fedora?

Probably nothing fedora-specific, though I'm not sure. It is *possible* the macro scriptlets - `Requires(pre|post|preun|postun)` - of the **systemd** and **shadow-utils** packages are fedora-specific.

**Comment by orbisvicis** [comment]
> We would ideally want to have our own rpm repo just like we do for debian, it would be great if you could give us some help with that.

Though I'm pretty new to fedora myself, I would:

00. Determine repository name and public uri

01. Get a fedora machine/vm

02. Install development environment and fedora build tools

        yum install @development-tools fedora-packager rpm-sign

03. Install repoview (optional) for repository web-ui

        yum install repoview

04. Create the directory containing the source build files

        for package in sonarr "${repository-name}-release"; do
            mkdir -p "/path/to/packages/${package}/"{RPMS,SRPMS,BUILD,SPECS,SOURCES}
        done

05. Create the directory containing the repository

        for release in 20 21; do
            mkdir -p "/path/to/repository/fedora/releases/${release}/"{source/SRPMS,noarch}
            mkdir -p "/path/to/repository/fedora/releases/${release}/"{i386,x86_64}/{os,debug}
        done

06. Create PGP signing key

        gpg --gen-key

07. Configure RPM package signing

        printf "%s\n" "%_signature gpg" >>~/.rpmmacros
        # make sure to get the right uid
        gpg --list-keys | sed -n 's#^uid:\s*\(.*\)#%_gpg_name \1#p' >>~/.rpmmacros

08. Create repository release package

        cd "/path/to/packages/${repository-name}-release"
        cd SOURCES

        gpg --export --armor "${GPG_ID}" >"RPM-GPG-KEY-${repository-name}"

        cat <<EOF >"${repository-name}.repo"
        # modelled after rpmfusion/livna
        ["${repository-name}"]
        name="${repository-name^}" for Fedora $releasever
        baseurl="${repository-uri}"'/fedora/releases/$releasever/$basearch/os/'
        enabled=1
        gpgcheck=1
        gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-${repository-name}-$releasever

        ["${repository-name}-debuginfo"]
        name="${repository-name^}" for Fedora $releasever - Debug
        baseurl="${repository-uri}"'/fedora/releases/$releasever/$basearch/debug/'
        enabled=0
        gpgcheck=1
        gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-${repository-name}-$releasever

        ["${repository-name}-source"]
        name="${repository-name^}" for Fedora $releasever - Source
        baseurl="${repository-uri}"'/fedora/releases/$releasever/source/SRPMS/'
        enabled=0
        gpgcheck=1
        gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-${repository-name}-$releasever
        EOF

        cd ../SPECS
        # modelled after rpmfusion/livna
        cat <<EOF >"${repository-name}-release.spec"
        Name:           ${repository-name}-release"
        Version:        1
        Release:        1
        Summary:        ${repository-name^} Repository Configuration

        License:        GPLv3
        URL:            ${upstream-url}
        Source1:        ${repository-name}.repo
        Source2:        RPM-GPG-KEY-${repository-name}
        BuildArch:      noarch


        %description
        <put description here>

        %prep
        echo "Nothing to prep"

        %build
        echo "Nothing to build"

        %install

        # Create dirs
        install -m 0755 -d \
          %{buildroot}%{_sysconfdir}/pki/rpm-gpg  \
          %{buildroot}%{_sysconfdir}/yum.repos.d

        # GPG Key
        install -m 0644 -p \
            %{SOURCE2} \
            %{buildroot}%{_sysconfdir}/pki/rpm-gpg

        # Yum .repo files
        install -m0644 -p \
            %{SOURCE1}
            %{buildroot}%{_sysconfdir}/yum.repos.d


        %files
        %{_sysconfdir}/pki/rpm-gpg/*
        %config(noreplace) %{_sysconfdir}/yum.repos.d/*

        %changelog
        * <date> <name> <email> - 1-1
        - <put changelog here>
        EOF

09. Configure mock (optional)

        cd /etc/mock
        # configure site-defaults.cfg if necessary
        # configure other files if necessary

10. create repository

        for release in 20, 21; do
            for arch in i386 x86_64 source noarch; do
                for version in os debug; do
                    path="/path/to/repository/fedora/release/${release}/${arch}/${version}"
                    createrepo "$path"
                    repoview "$path"
                    gpg --detach-sign --armor "${path}/repodata/repomd.xml"
                done
            done
        done
11. build and install new/modified packages

        packages=(sonarr "${repository-name}-release")
        for release in 20 21; do
            for arch in x86_64 i386; do
                chroot="fedora-${release}-${arch}"
                for package in "${packages[@]}"; do
                    pathsrc="/path/to/packages/${package}"
                    pathrpo="/path/to/repository/fedora/release/${release}"
                    pathmck="${mock-basedir}/${chroot}/result"
                    cd "$pathsrc"
                    spectool --directory "${pathsrc}/SOURCES" "${pathsrc}/SPECS/"*.spec
                    # couldn't figure out how to properly separate the
                    # srpm/debuginfo/regular rpm packages
                    mock \
                        --root "$chroot" \
                        --buildsrpm \
                        --sources "${pathsrc}/SOURCES" \
                        --spec "${pathsrc}/SPECS/"*.spec \
                    mock \
                        --root "$chroot" \
                        --rebuild "${pathmck}/"*.src.rpm
                    rpm --addsign "${pathmck}/"*.rpm
                    # hopefully no packages are named "*debuginfo*"
                    mv "${pathmck}/"*.src.rpm "${pathrpo}/source/SRPMS"
                    mv "${pathmck}/"*debuginfo*.rpm "${pathrpo}/${arch}/debug"
                    mv "${pathmck}/"*.rpm "${pathrpo}/${arch}/os"
                done
            done
        done

12. update repository

        for release in 20, 21; do
            for arch in i386 x86_64 source noarch; do
                for version in os debug; do
                    path="/path/to/repository/fedora/release/${release}/${arch}/${version}"
                    createrepo --update "$path"
                    repoview "$path"
                    gpg --detach-sign --armor "${path}/repodata/repomd.xml"
                done
            done
        done

13. Instruct users

        su -c 'yum localinstall --nogpgcheck "${repository-uri}"/fedora/releases/$(rpm -E %fedora)/noarch/os/"${repository-name}-release-${version}-${release}".noarch.rpm'

Of course, the repository has to be made public somehow, and you'll probably want to integrate build-failure detection in the final script.


Resources:

* http://yum.baseurl.org/wiki/RepoCreate
* http://linux.die.net/man/8/createrepo
* http://linux.die.net/man/8/repoview
* https://fedoraproject.org/wiki/How_to_create_an_RPM_package
* https://fedoraproject.org/wiki/Packaging:RPMMacros?rd=Packaging/RPMMacros
* http://www.rpm.org/max-rpm-snapshot/s1-rpm-specref-files-list-directives.html
* https://fedoraproject.org/wiki/Projects/Mock
* https://wiki.archlinux.org/index.php/GnuPG
* http://www.rpm.org/max-rpm/s1-rpm-pgp-signing-packages.html
* http://www.redhat.com/archives/rhl-list/2007-August/msg04261.html
* http://linuxsysconfig.com/2013/04/create-a-yum-repository-with-custom-gpg-signed-packages/
* https://iuscommunity.org/pages/ConfiguringADedicatedRPMDevelopmentBox.html
* https://iuscommunity.org/pages/CreatingAGPGKeyandSigningRPMs.html
* http://www.openfusion.net/linux/mocking_rpms
* http://rpmfusion.org/Configuration
* http://download1.rpmfusion.org/free/fedora/releases/
* http://rpm.livna.org/repo/

File: fedora/sonarr.systemd [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+Documentation=https://sonarr.tv/
+Wants=network.target
+After=network.target
+
+
+[Service]
+Type=simple
+User=sonarr
+PIDFile=/var/lib/sonarr/.config/NzbDrone/nzbdrone.pid
+ExecStart=/usr/bin/mono /opt/sonarr/NzbDrone.exe
```
Comment (Taloth) [inline fedora/sonarr.systemd]: --debug to get lines in logs?

**Comment by halkeye** [comment]
I don't know if its very applicable in this case (haven't actually used it yet) but have you looked at fpm yet?

**Pull request closed without merging**